- Since this data have been scraped from the web I am assuming there will be some manual cleaning necessary before I can perform exploratory data analysis on the dataset. Below are the steps that I took in doing so:

In [81]:
import numpy as np
import pandas as pd

# Loading the Data:

In [23]:
beers = pd.read_csv('untapped_beer_data_eng_language.csv')

In [24]:
beers.head()

name              brewery beer_style       abv  date_added  \
0       Exalted  Thornbridge Brewery    Altbier  4.5% ABV  06/10/2016   
1          Leaf   Orbit Beers London    Altbier  6.2% ABV  12/03/2014   
2  Ctrl Alt Del                Wylam    Altbier  4.9% ABV  02/07/2016   
3         Hades        Ghost Brew Co    Altbier  4.5% ABV    07/16/16   
4   The Altbier     Anspach & Hobday    Altbier    5% ABV  10/02/2015   

   not_available num_ratings  rating  \
0            NaN         542    3.70   
1            NaN         783    3.67   
2            NaN         318    3.65   
3            NaN         228    3.62   
4            NaN         288    3.57   

                                           beer_desc  
0                      Collaboration with Magic Rock  
1  Beechwood smoked malt meets a double altbier.\...  
2  12.4° Plato | OG 1050 | ABV 4.9%\n\nWylam & Le...  
3                                Dry Hopped Altbier.  
4                                                NaN

- Looking at the size of the dataframe.
- This checks I have enough data to analyse and also is a good check post-cleaning to see how many datapoints cleaning may end up removing.

In [25]:
beers.shape

(22984, 9)

# Looking at the Variables:

## Number of unique attributes per variable:

In [26]:
beers.nunique(axis=0)

name             21511
brewery           3892
beer_style         213
abv                305
date_added        3486
not_available        1
num_ratings       4666
rating             235
beer_desc        19332
dtype: int64

- As we can see, because of the categorical variables this dataset exhibits high cardinality.

## Description of the numerical columns:

In [27]:
beers.describe()

not_available        rating
count         4357.0  22984.000000
mean             1.0      3.779146
std              0.0      0.260250
min              1.0      1.790000
25%              1.0      3.620000
50%              1.0      3.750000
75%              1.0      3.910000
max              1.0      4.740000

- Right now there are only two numerical variables in the dataframe, although I'd expect four (not available, rating, ABV and number of ratings). 
- ABV and number of ratings are being recognised as a non-numeric type, I will rectifty this below.

## A look at the null values included in the dataframe:

In [28]:
beers.isnull().sum()

name                 0
brewery              0
beer_style           0
abv                  0
date_added           0
not_available    18627
num_ratings          0
rating               0
beer_desc         3481
dtype: int64

- Not available indicates whether a beer has been discontinued (i.e. is not available for purchase).

- At present a 1.0 indicates 'not available' and a NaN indicates it is still purchasable.

- I will fill the NaN in this column with zero's and then change the column so it more easily understood.

In [29]:
# Replacing NaNs with zero's:
beers['not_available'] = beers['not_available'].fillna(0)

In [30]:
beers.not_available.value_counts()

0.0    18627
1.0     4357
Name: not_available, dtype: int64

In [31]:
# Swapping 1 for 0 in the not_available column
beers.not_available = beers.not_available.map( {0:1 , 1:0} )

# Changing the column name to be called available - so 1's now equate to a beer that is purchasable.
beers = beers.rename(columns={'not_available': 'available'})

In [32]:
beers.available.value_counts()

1    18627
0     4357
Name: available, dtype: int64

### Filling the beer_desc null values with zero for now:

- These are beers that either don't have descriptions on the can or the descriptions have not been uploaded to untapped, during EDA I will inspect what % of my dataset has no description and decide what to do about that.

In [33]:
beers['beer_desc'] = beers['beer_desc'].fillna(0)

In [34]:
beers.isnull().sum()

name           0
brewery        0
beer_style     0
abv            0
date_added     0
available      0
num_ratings    0
rating         0
beer_desc      0
dtype: int64

## Looking at the datatypes of the columns:

In [35]:
beers.dtypes

name            object
brewery         object
beer_style      object
abv             object
date_added      object
available        int64
num_ratings     object
rating         float64
beer_desc       object
dtype: object

### Inspecting why num_ratings is an object datatype:

- num_ratings is the number of ratings that the beer has received, so it should be an int type.

In [36]:
# Looking at instances where num_ratings is not numeric:
beers[beers['num_ratings'].str.isnumeric() == False]

name                       brewery     beer_style  \
8                       Neu            Orbit Beers London        Altbier   
10               London Alt        Mondo Brewing Company         Altbier   
19             DoppelSticky     Lagunitas Brewing Company        Altbier   
30                Altbitter       Widmer Brothers Brewing        Altbier   
39               Sticke Alt   Lion Bridge Brewing Company        Altbier   
...                     ...                           ...            ...   
22970            Holly King             Green Man Brewery  Winter Warmer   
22972  High Camp Winter IPA  Bale Breaker Brewing Company  Winter Warmer   
22973       Marley's Lament      Alewerks Brewing Company  Winter Warmer   
22980            Zoigl-Pils     Zoiglhaus Brewing Company          Zoigl   
22983      Zoigl Star Lager   Appalachian Brewing Company          Zoigl   

            abv  date_added  available num_ratings  rating  \
8      4.7% ABV    08/26/14          0       1,780    3.47   
10     4.8% ABV    02/24/15          1       1,227    3.45   
19     7.7% ABV  09/10/2015          1      12,203    3.87   
30     9.5% ABV  09/08/2014          1       2,276    3.78   
39       6% ABV    04/15/14          1       1,054    3.75   
...         ...         ...        ...         ...     ...   
22970   11% ABV  12/03/2013          1       2,302    3.81   
22972  7.3% ABV    11/22/13          1       2,524    3.81   
22973    8% ABV    11/30/13          1       2,137    3.80   
22980  4.8% ABV    03/16/17          1       2,188    3.74   
22983  5.8% ABV    08/21/10          1       1,562    3.47   

                                               beer_desc  
8      Pours deep amber colour with beautiful reddish...  
10     The London Alt Beer is a original Dusseldorf-s...  
19     This fruity, estery double altbier is brewed i...  
30     AltBitter, a collaboration beer between Deschu...  
39     A traditional German Ale that has a pleasing b...  
...                                                  ...  
22970  This rich holiday ale is perfect to warm away ...  
22972  Named for the ski lodge atop our local White P...  
22973  This winter warmer is perfect for a cold day. ...  
22980  2017 Great American Beer Festival Gold Medal i...  
22983  Zoigl is a golden, unfiltered lager beer which...  

[6909 rows x 9 columns]

- Having looked at the csv file, there are commas, spaces and the word 'Ratings' preventing the num_ratings column being typed as an integer. Below I have removed these values from this column:

In [37]:
beers.num_ratings = [int(x.strip(' ').replace(',', '').replace('Ratings', '')) for x in beers.num_ratings]

In [38]:
# num_ratings now changed to an int column:
beers.dtypes

name            object
brewery         object
beer_style      object
abv             object
date_added      object
available        int64
num_ratings      int64
rating         float64
beer_desc       object
dtype: object

### Removing % ABV from abv column to convert it to float:

In [41]:
# Removing the spaces, % signs and ABV from column:
beers.abv = [(x.strip(' ').replace('%', '').replace('ABV', '')) for x in beers.abv]

In [42]:
beers.head()

name              brewery beer_style   abv  date_added  available  \
0       Exalted  Thornbridge Brewery    Altbier  4.5   06/10/2016          1   
1          Leaf   Orbit Beers London    Altbier  6.2   12/03/2014          1   
2  Ctrl Alt Del                Wylam    Altbier  4.9   02/07/2016          1   
3         Hades        Ghost Brew Co    Altbier  4.5     07/16/16          1   
4   The Altbier     Anspach & Hobday    Altbier    5   10/02/2015          1   

   num_ratings  rating                                          beer_desc  
0          542    3.70                      Collaboration with Magic Rock  
1          783    3.67  Beechwood smoked malt meets a double altbier.\...  
2          318    3.65  12.4° Plato | OG 1050 | ABV 4.9%\n\nWylam & Le...  
3          228    3.62                                Dry Hopped Altbier.  
4          288    3.57                                                  0

- There are still some columns with N/A in the abv:

In [59]:
beers[27:28]

name                       brewery beer_style   abv  \
27  Bourbon Barrel Aged Copper  The Olde Mecklenburg Brewery    Altbier  N/A    

    date_added  available  num_ratings  rating beer_desc  
27  09/11/2016          1          173    3.79         0

In [63]:
beers[beers['abv']=='N/A ']

name  \
27                      Bourbon Barrel Aged Copper   
42             Bourbon Barrel Aged Anniversary Alt   
45                                 Anniversary Alt   
111                         Civil Disobedience #14   
115                               Elaborative Five   
...                                            ...   
22277                                       Uberon   
22393                                Peach Crumble   
22427                Candy Floss Pomegranate Wheat   
22656                          Naked for Satan Ale   
22827  Barrel Aged 12 Dogs of Christmas Ale (2015)   

                            brewery                        beer_style   abv  \
27     The Olde Mecklenburg Brewery                           Altbier  N/A    
42     The Olde Mecklenburg Brewery                           Altbier  N/A    
45     The Olde Mecklenburg Brewery                           Altbier  N/A    
111          Hill Farmstead Brewery                 American Wild Ale  N/A    
115          Hill Farmstead Brewery                 American Wild Ale  N/A    
...                             ...                               ...   ...   
22277                Bell's Brewery  Wheat Beer - American Pale Wheat  N/A    
22393            Lil Beaver Brewery                Wheat Beer - Other  N/A    
22427    Heavy Riff Brewing Company              Wheat Beer - Witbier  N/A    
22656   Matilda Bay Brewing Company              Wheat Beer - Witbier  N/A    
22827   Thirsty Dog Brewing Company                        Winter Ale  N/A    

       date_added  available  num_ratings  rating  \
27     09/11/2016          1          173    3.79   
42     10/04/2014          1          163    3.74   
45     03/08/2013          1         1047    3.74   
111    07/09/2015          0         1402    4.44   
115    08/09/2016          0         2090    4.42   
...           ...        ...          ...     ...   
22277    06/16/13          1         7015    3.93   
22393  03/03/2020          1          637    3.86   
22427    09/24/16          1          302    3.77   
22656    11/16/11          1          258    3.51   
22827  12/04/2015          0          240    3.91   

                                               beer_desc  
27                                                     0  
42                                                     0  
45     Anniversary Alt is brewed every March for our ...  
111    A blend of various beers/barrels from Summer 2...  
115    100% Spontaneously Fermented/Coolshipped beer ...  
...                                                  ...  
22277  Barrel-aging this beer coaxes flavors out of T...  
22393         Wheat Ale with Peach, Cinnamon, and Vanill  
22427  This pomegranate wheat is one of our originals...  
22656                                                  0  
22827                         Bourbon barrel aged 12 dog  

[202 rows x 9 columns]

- Since these are only 202 n/a values out of a dataframe of 20,000 values I am happy to drop N/A abv rows as I think abv is an important feature, and filling with either zero's, mean, median or knn would not be very meaningful.

In [66]:
# Removing further spaces and changing N/A string to 0 string to convert to numeric later
beers.abv = [(x.strip(' ').replace('N/A', '0')) for x in beers.abv]

In [69]:
beers[27:28]

name                       brewery beer_style abv  \
27  Bourbon Barrel Aged Copper  The Olde Mecklenburg Brewery    Altbier   0   

    date_added  available  num_ratings  rating beer_desc  
27  09/11/2016          1          173    3.79         0

In [78]:
beers.head()

name              brewery beer_style  abv date_added  available  \
0       Exalted  Thornbridge Brewery    Altbier  4.5 2016-06-10          1   
1          Leaf   Orbit Beers London    Altbier  6.2 2014-12-03          1   
2  Ctrl Alt Del                Wylam    Altbier  4.9 2016-02-07          1   
3         Hades        Ghost Brew Co    Altbier  4.5 2016-07-16          1   
4   The Altbier     Anspach & Hobday    Altbier  5.0 2015-10-02          1   

   num_ratings  rating                                          beer_desc  
0          542    3.70                      Collaboration with Magic Rock  
1          783    3.67  Beechwood smoked malt meets a double altbier.\...  
2          318    3.65  12.4° Plato | OG 1050 | ABV 4.9%\n\nWylam & Le...  
3          228    3.62                                Dry Hopped Altbier.  
4          288    3.57                                                  0

In [71]:
# Changing column to a numeric column:
beers['abv'] = pd.to_numeric(beers['abv'])

In [74]:
beers.shape

(22984, 9)

In [75]:
# Dropping any rows in abv that are equal to zero:
beers = beers[beers['abv'] != 0]

In [76]:
beers.shape  #202 rows removed

(22782, 9)

In [73]:
# abv now changed to an float column:
beers.dtypes

name            object
brewery         object
beer_style      object
abv            float64
date_added      object
available        int64
num_ratings      int64
rating         float64
beer_desc       object
dtype: object

### Converting date_added column to datetime:

In [77]:
beers['date_added'] =  pd.to_datetime(beers['date_added'])

In [79]:
beers.dtypes

name                   object
brewery                object
beer_style             object
abv                   float64
date_added     datetime64[ns]
available               int64
num_ratings             int64
rating                float64
beer_desc              object
dtype: object

# Creating a new column for Amalgamated Beer Styles:

- There are 213 (see below) unique beer styles included in the dataset, with some being very similar to others (e.g. 'Lager - Amber', 'Lager - American', 'Lager - American Amber / Red', 'Lager - American Light').
- I will create a new column where all IPA's, all lagers, all pale ales etc are grouped together.

In [93]:
beers.beer_style.nunique()

213

In [94]:
# Creating a function to match words in the beer_style column to those I've listed as popular beers:
# Popular beers are those that have more than one variant - i.e. Bock Doppelbok and Bock - Eisbock or Brown Ale - English and 
# Brown Ale - American

def extract_beer_supgroup(beer):
    pop_beer =  ['Barleywine', 'Belgian', 'Bock', 'Brown Ale', 'Cider', 'Farmhouse Ale', 'IPA', 'Lager', 
                'Lambic', 'Mead', 'Pale Ale', 'Pilsner', 'Porter', 'Red Ale', 'Sour', 'Stout', 'Strong Ale', 'Wheat Beer']
    beer_supgroup = beer.split('-')[0].strip(' ')
    if beer_supgroup in pop_beer:
        supgroup = beer_supgroup
    elif beer.split(' ')[0] == 'Belgian':
        supgroup = 'Belgian'
    else:
        supgroup = 'Other'
        
    return supgroup   

In [95]:
beers['beer_style_group'] = [extract_beer_supgroup(x) for x in beers.beer_style]

In [97]:
# Checking that my function has executed correctly:
pd.set_option('display.max_rows', 230)
beers[['beer_style', 'beer_style_group']].drop_duplicates()

beer_style beer_style_group
0                                  Altbier            Other
74                       American Wild Ale            Other
152               Australian Sparkling Ale            Other
155                             Golden Ale            Other
191                  Barleywine - American       Barleywine
298                   Barleywine - English       Barleywine
425                     Barleywine - Other       Barleywine
501                         Belgian Blonde          Belgian
601                         Belgian Dubbel          Belgian
704                      Belgian Quadrupel          Belgian
803                Belgian Strong Dark Ale          Belgian
810                         Belgian Tripel          Belgian
894              Belgian Strong Golden Ale          Belgian
1087       Bière de Champagne / Bière Brut            Other
1146                           Black & Tan            Other
1191                            Blonde Ale            Other
1404                     Bock - Doppelbock             Bock
1483          Bock - Eisbock (Traditional)             Bock
1526     Bock - Hell / Maibock / Lentebock             Bock
1588           Bock - Single / Traditional             Bock
1666                     Bock - Weizenbock             Bock
1730               Bock - Weizendoppelbock             Bock
1742                            Hefeweizen            Other
1745                            Brett Beer            Other
1756                  Brown Ale - American        Brown Ale
1900                   Brown Ale - English        Brown Ale
1920                   Brown Ale - Belgian        Brown Ale
2140         Brown Ale - Imperial / Double        Brown Ale
2231                     Brown Ale - Other        Brown Ale
2353                            Burton Ale            Other
2385                     California Common            Other
2494                   Chilli / Chile Beer            Other
2500                  Spiced / Herbed Beer            Other
2514                           Cider - Dry            Cider
2625                         Cider - Graff            Cider
2645      Cider - Herbed / Spiced / Hopped            Cider
2659                   Cider - Traditional            Cider
2731               Cider - Ice / Applewine            Cider
2753                   Cider - Other Fruit            Cider
2865                         Cider - Perry            Cider
2982                          Cider - Rosé            Cider
3039                         Cider - Sweet            Cider
3358                             Cream Ale            Other
3463                             Dampfbier            Other
3501                              Dark Ale            Other
3664                          Dunkelweizen            Other
3739                        English Bitter            Other
3791                         IPA - English              IPA
3971                      English Mild Ale            Other
4106         Extra Special / Strong Bitter            Other
4257        Farmhouse Ale - Bière de Garde    Farmhouse Ale
4333         Farmhouse Ale - Bière de Mars    Farmhouse Ale
4384                 Farmhouse Ale - Other    Farmhouse Ale
4471                 Farmhouse Ale - Sahti    Farmhouse Ale
4525                Farmhouse Ale - Saison    Farmhouse Ale
4782                              Festbier            Other
4845                 Freeze-Distilled Beer            Other
4858                            Fruit Beer            Other
4996                           Ginger Beer            Other
5128                           Gluten-Free            Other
5533                   Grape Ale - Italian            Other
5536                     Grape Ale - Other            Other
5559                              Grisette            Other
5581                          Sour - Other             Sour
5656            Gruit / Ancient Herbed Ale            Other
5697                        Sour - Fruited             Sour
5720        

In [98]:
beers['beer_style_group'].value_counts()

Other            5596
IPA              3663
Stout            2506
Lager            1615
Pale Ale         1441
Sour             1292
Porter            990
Cider             841
Red Ale           622
Brown Ale         593
Belgian           587
Pilsner           543
Farmhouse Ale     515
Wheat Beer        485
Mead              358
Bock              338
Barleywine        310
Strong Ale        245
Lambic            242
Name: beer_style_group, dtype: int64

# Saving resulting Dataframe to csv file:

In [99]:
beers.to_csv('beers_data_cleaned.csv')